accord-core/src/main/java/accord/local/Command.java

```diff
@@ -399,158 +525,253 @@ private void onChangeInternal(PartialCommand command)
-                if (notifyListenersOnNoop) notifyListeners();
-                return Writes.SUCCESS;
-            }
-            assert !isWaitingOnApply();
+            if (alwaysNotifyListeners)
+                notifyListeners(safeStore);
+
+            if (notifyWaitingOn)
+                new NotifyWaitingOn(this).accept(safeStore);
+            return false;
         }
 
         switch (status())
         {
             case Committed:
                 // TODO: maintain distinct ReadyToRead and ReadyToWrite states
-                status(ReadyToExecute);
+                setStatus(ReadyToExecute);
                 logger.trace("{}: set to ReadyToExecute", txnId());
-                boolean isProgressShard = progressKey() != null && handles(txnId().epoch, progressKey());
-                commandStore().progressLog().readyToExecute(this, isProgressShard, isProgressShard && progressKey().equals(homeKey()));
-                notifyListeners();
+                safeStore.progressLog().readyToExecute(this, shard);
+                notifyListeners(safeStore);
                 break;
-            case Executed:
-                logger.trace("{}: applying", txnId());
-                if (notifyListenersOnNoop) notifyListeners();
-                return apply();
+
+            case PreApplied:
+                if (executeRanges(safeStore, executeAt()).intersects(writes().keys, safeStore.commandStore()::hashIntersects))
+                {
+                    logger.trace("{}: applying", txnId());
+                    apply(safeStore);
+                }
+                else
+                {
+                    logger.trace("{}: applying no-op", txnId());
+                    setStatus(Applied);
+                    notifyListeners(safeStore);
+                }
         }
-        return Writes.SUCCESS;
+        return true;
     }
 
     /**
      * @param dependency is either committed or invalidated
+     * @param isInsert true iff this is initial {@code populateWaitingOn} call
+     * @return true iff {@code maybeExecute} might now have a different outcome
      */
-    private void updatePredecessor(PartialCommand dependency)
+    private boolean updatePredecessor(Command dependency, boolean isInsert)
     {
         Preconditions.checkState(dependency.hasBeen(Committed));
         if (dependency.hasBeen(Invalidated))
         {
             logger.trace("{}: {} is invalidated. Stop listening and removing from waiting on commit set.", txnId(), dependency.txnId());
             dependency.removeListener(this);
-            removeWaitingOnCommit(dependency);
+            removeWaitingOnCommit(dependency.txnId()); // TODO (now): this was missing in partial-replication; might be redundant?
+            return true;
         }
         else if (dependency.executeAt().compareTo(executeAt()) > 0)
         {
             // cannot be a predecessor if we execute later
```

Comment:
```suggestion
            // cannot be a predecessor if they execute later
```
Log message says different

Reply:
Depends how you interpret `we` and `they`. I will update to make it unambiguous.